### DISCUSSION TITLE
动态创建的PAGTextLayer 如何设置坐标?

### DISCUSSION THREAD

**Discussion by cgeffect** [Tencent/libpag#2225] [General] [answered]
在PAGPlayer函数中可以给PAGPlayer设置PAGComposition
void setComposition(std::shared_ptr<PAGComposition> newComposition);

PAGComposition又可以通过static std::shared_ptr<PAGComposition> Make(int width, int height);函数创建, 
在PAGComposition的成员函数中可以通过  bool addLayer(std::shared_ptr<PAGLayer> pagLayer);函数增加一个图层

伪代码

auto comPtr = std::shared_ptr<PAGComposition> Make(100, 100);
std::shared_ptr<TextDocument> textDocumentHandle;
auto textPtr = std::shared_ptr<PAGTextLayer> Make(10000,textDocumentHandle);
comPtr->addLayer(textPtr);
auto pagPlayer = new pag::PAGPlayer();
pagPlayer->setComposition(comPtr);

1. 可以在没有PAG文件的情况下创建一个PAGComposition吗, 然后赋值给PAGPLayer, 进行渲染.
2. 被添加的PAGTextLayer相对于PAGComposition的xy坐标是在哪里, 没有看到设置位置的函数.

**Comment by kevingpqi123** [comment, marked as answer]
1、PAGComposition 就是一个容器，支持自己创建，具体可以看下 API 文档：
2、对于图形绘制而言，设置位置都是通过 setMatrix，相关用法可以网上搜索下或看下源码中 pag 的测试用例

**Reply by cgeffect** [reply]
[4.3版本](tag: v4.3.51, origin/release/4.3)
请问下在 mac 上 PAGTextLayer objective-c 的接口没有 对应的C++的Make接口. 设计如此吗?
std::shared_ptr<PAGTextLayer> PAGTextLayer::Make(int64_t duration, std::string text, float fontSize, std::string fontFamily, std::string fontStyle)

**Reply by cgeffect** [reply]
在linux的官方demo程序中, 自己添加一个PAGTextLayer层, 在未设置字体大小和设置字体大小为100时渲染结果如下

![image](https://github.com/Tencent/libpag/assets/15692322/0ebe0926-edd5-4f1d-88f7-b55408ea20a1)

代码如下:  
std::shared_ptr<pag::PAGFile> ReplaceImageOrText() {
  auto pagFile = pag::PAGFile::Load("../../assets/720.pag");
  if (pagFile == nullptr) {
    return nullptr;
  }
  
  {
    std::shared_ptr<pag::TextDocument> textDocumentHandle = std::make_shared<pag::TextDocument>();
    // textDocumentHandle->baselineShift = -105;
    textDocumentHandle->applyStroke = true;
    textDocumentHandle->strokeWidth = 4;
    textDocumentHandle->strokeOverFill = false;
    textLayer = pag::PAGTextLayer::Make(1000 * 1000, textDocumentHandle);
    pagFile->addLayerAt(textLayer, 1);
    textLayer->setText("默认文本");
    pag::Color col = {205, 179, 221};
    textLayer->setFillColor(col);
    textLayer->setStrokeColor(pag::White);
    textLayer->setFontSize(100);
    pag::Rect rect = textLayer->getBounds();
    std::cout << rect.left << " " << rect.right << " " << rect.top << " " << rect.bottom << std::endl;
  }

  return pagFile;
}

为什么设置成100时, 文字绘制会向上. 如果需要让文字对齐左上角显示, Y轴应该往下移动, 移动多少呢? 这个值要怎么计算?

**Comment by kevingpqi123** [comment]
可以参考下 iOS 平台的这个 demo， https://github.com/libpag/pag-ios/blob/master/sample/pag-ios/ViewController/MoveAndRotateLayerViewController.swift
